src/util/vec2d.rs

```diff
@@ -0,0 +1,160 @@
+use std::ops::{Index, IndexMut};
+
+/// A 2 dimensional matrix in row-major order backed by a contiguous slice.
+#[derive(Debug)]
+pub struct Vec2D<T> {
+    data: Box<[T]>,
+    cols: usize,
+}
+
+impl<T> Vec2D<T> {
+    /// Initialize a grid of size (`rows`, `cols`) with the given data element.
+    pub fn init(data: T, size: (usize, usize)) -> Vec2D<T>
+    where
+        T: Clone,
+    {
+        let (rows, cols) = size;
+        let len = rows
+            .checked_mul(cols)
+            .unwrap_or_else(|| panic!("{} rows by {} cols exceeds usize::MAX", rows, cols));
+        Vec2D {
+            data: vec![data; len].into_boxed_slice(),
+            cols,
+        }
+    }
+
+    /// Fills the grid with elements by cloning `value`.
+    pub fn fill(&mut self, value: T)
+    where
+        T: Clone,
+    {
+        self.data.fill(value)
+    }
+}
+
+impl<T> Index<usize> for Vec2D<T> {
+    type Output = [T];
+
+    #[inline]
+    fn index(&self, row: usize) -> &Self::Output {
+        let start_row = row * self.cols;
+        &self.data[start_row..start_row + self.cols]
```

Comment:
Here we don't explicitly check that `start_row < rows`, so this addition may overflow, for example:

```rust
// Matrix with 5 columns.
let matrix = Vec2D::init(0, (3, 5));
// Somehow, as long as numbits(usize) is a multiple of 4, then 5 divides usize::MAX.
// This is clear in hexadecimal: usize::MAX is 0xFFF...F and usize::MAX / 5 is 0x333...3.
let row = usize::MAX / 5;
// This will therefore try to index data[usize::MAX..4].
matrix[row];
```

In practice this looks to be already caught by Rust's `std::slice::index`, but please add a `should_panic` test for this as well.

Reply:
Added a test for this which panics correctly, but did not bother with `checked_add`.